### DISCUSSION TITLE
Add a code of conduct

### DISCUSSION THREAD

**Discussion by cessen** [helix-editor/helix#5585] [General]
Many prominent Rust projects, including Rust itself, have a code of conduct that explicitly outlines the expected behavior while participating in the project.  I don't think we've had any issues so far, and I suspect we won't have issues in practice that can't be handled on an ad-hoc basis, at least not for quite some time.

Nevertheless, outlining behavioral expectations explicitly can make some people feel safer in approaching the project, and generally create a more welcoming environment.  It sort of serves as a sign post that says, "Hey, we have an explicit goal to be reasonable and friendly here".

Rust itself uses the [Contributor Covenant](https://www.contributor-covenant.org/version/1/3/0/code-of-conduct/), which would serve as a reasonable basis, I think.  However, as-is I think it comes across to many people as code for "this is a politically left-leaning project".  That may actually make the project *less* welcoming to people that don't share those political leanings, and I don't think that's a message we want to send.  (And I say that as someone who is pretty far left-leaning myself.)

Another option to use as a basis is [Microsoft's open source code of conduct](https://microsoft.github.io/codeofconduct/).  It serves the same purpose as the Contributor Covenant and covers pretty much the same things, but feels more politically neutral.  It even explicitly lists political beliefs as a protected trait (which is somewhat conspicuously missing from the Contributor Covenant).

In either case we'd need to adapt things a bit.

Thoughts?

**Comment by kirawi** [comment]
They both sound the same to me, to be honest. Microsoft's just sounds more like legalese while the first one sounds like an active commitment the project is making to its users. I prefer Rust's.

**Comment by cessen** [comment]
Ah, to be clear, my preference for Microsoft's code of conduct is mainly with respect to the top section about expected behavior and community aspirations.  It feels more politically neutral to me, as well as covering not only prohibited behaviors but also positive behaviors to aspire to.  Just over-all it feels more balanced to me.

The sections about how issues are reported and handled I feel less strongly about, and I agree that those sections in the Microsoft code of conduct are pretty dry and legalese-like.

**Comment by kirawi** [comment]
I believe that the Contributor Covenant is winning me over because it has a more active voice compared to Microsoft's. "We will not tolerate so and so" versus "So and so are not tolerated". I don't have a preference when it comes to the actual contents, but I find that an active voice is more effective in leaving a lasting impression on the reader and encouraging them to follow it. If Microsoft's CoC could be rewritten in active voice, that would also work for me.

**Comment by pickfire** [comment]
Maybe just use rust code of conduct. I think rust one is quite good since many people is happy with rust. I don't know whether if it's left leaning or right leaning but if someone did make others feel bad that will possibly reduce the happiness of the contributors, it's open source so each contributors counts.

For me both works, but I think rust is better. Any is fine for me, I think most people here right now.

**Comment by cessen** [comment]
So, apparently I misremembered.  Rust doesn't use the Contributor Covenant, they instead rolled their own code of conduct, pulling ideas from various places (including the Covenant):
https://www.rust-lang.org/policies/code-of-conduct

In particular, they also highlight positive community aspirations as well, which I appreciate.

In general, I'm starting to feel like rolling our own as well probably makes the most sense...?  A code of conduct is ultimately about community values, and making those values explicit.

And the more I think about it, the more I'm realizing the main thing that bugs me about the Contributor Covenant is that (IMO) it's a poor statement of values.  It's really more a list of rules and consequences.  And while explicitly outlining unacceptable behavior is certainly a good thing to do, and I also think we should do that, I'd ideally like the code of conduct here to be more balanced.  Less fire-and-brimstone and more "here's the kind of community we want to be".

**Comment by kirawi** [comment]
Maybe this should be converted into a discussion?

**Comment by cessen** [comment]
I don't really think that's necessary.  This issue already serves as a good place for discussion.  And when we're ready to get more concrete with some actual proposed text, a PR probably makes the most sense, so people can provide feedback on specific parts of it.

**Comment by jficz** [comment]
I've seen a very short "community guideline" somewhere which only said:
> Basically, just don't be an asshole.

That covers everything imo. The more complex it gets, the more "political" it gets and the more trouble with it there will be. One can never accommodate everyone and imho any community guideline that goes beyond "don't be an asshole" is only inviting people who will probably have a problem with the guidelines one way or another. In a technical project you probably don't want to spend too much time playing arbiter between people who think the guidelines are too strict and people who think they're not strict enough.

**Comment by kirawi** [comment]
Personally, I think that leaves too much at the discretion of the reviewer. Having a defined set rules, even if they're flawed, eliminates ambiguity.

**Comment by dead10ck** [comment]
Honestly that's so vague and open to interpretation that it's just satirical of code of conducts in general.

**Comment by cessen** [comment]
> Basically, just don't be an asshole.

I mean, it's not a bad starting point.  The issue is not everyone agrees on the specifics of what that means, so even well-meaning people may run afoul of community norms around that.

There's also an aspect of CoCs that can act as a signal to minority groups that, "Hey, we are explicitly trying to be welcoming to you as well."  Which I think is worthwhile.

> The more complex it gets, the more "political" it gets

I don't think the complexity is necessarily what makes it political, but I agree that a lot of CoCs are left leaning (possibly without realizing it) in a way that can make people with other political leanings feel unwelcome, and I think that's bad as well.  IMO a good CoC should aspire to [signal a welcoming environment for everyone who's willing to uphold that welcoming environment themselves](https://blog.cessen.com/post/2018_09_18_politics_and_open_source_communities).
